### DISCUSSION TITLE
Ingredient sections

### DISCUSSION THREAD

**Discussion by nr0dham** [TomBursch/kitchenowl#160] [Ideas]
I'm pretty new but loving KitchenOwl so far!

I've started putting in recipes, and already noticed some dynamics regarding ingredients (items).

Some recipes have sections of ingredients, such as for the main part, and then a sauce to put on top, or a crust. Then, the directions specify "to make the sauce, combine ingredients in a bowl", etc.

Given that there's only one list of ingredients available for the recipe, I have to list out the "sauce" ingredients, like "combine the cream, butter, flour, etc", and then when preparing, I think it's a bit unwieldy to find all of those in the list of recipe items.

Also, some recipes have, for example, "butter" more than once - one, melted butter for a sauce, regular butter as a main dish ingredient.  I don't think it's currently possible to have two "butter"s, but with ingredient sections, it could be.

So: I'm thinking it would be good to have a way to create "sections" of ingredients, like "sauce", "main dish", etc, within a given recipe.  That would require a button to add a section, and a way to associate an item being added to a section.

Does that resonate with anybody else?  Should I figure out UI and put in a request?

Edit: Just realized perhaps the sections could be called "recipe components"?  Like "sauce, crust, filling" as "components"?

**Comment by TomBursch** [comment]
Thanks for the idea! I noticed the same, but only in a couple of my recipes, never enough to come up with a solution. I like the idea of splitting the recipes in parts/components somehow. Internally, these parts could be represented as new recipes and then linked to the main one. Not sure if these sub-recipes should be part of the household and shown in the recipe list or be hidden. 

If you can, feel free to propose a UI for it, always happy for suggestions and inspirations.

**Reply by nr0dham** [reply]
Thanks!  I was thinking it could be done by adding a field "component_id" to the "recipe_items" table, then adding a table called "components" or whatever with the recipe id and name of the component.  When you list out the items, if there are any that are component 1 or 2 or whatever instead of 0, look up the name, print a header, iterate through the items, repeat for each component.

I'll work on UI, have to noodle that a little.

**Reply by nr0dham** [reply]
Oh hey, or maybe just something much like tagging, but within the recipe...

**Reply by nr0dham** [reply]
Yep, thinking about it, I think the "sub recipe" thing is a bit ambitious personally.  Just having the ingredients in sections is good enough.  I feel like, when adding an item, just have an interface just like the recipe tags, but put "Component:" before it, with the "+" to add a "component tag", and that's the UI.  When you're listing out the items, do groups for component 1, component 2, etc, with name headers, THEN do component 0, which is untagged items.  Done.  How's that sound?

**Reply by irishrain** [reply]
I like the idea of sub recipies. You could even allow reusing these sub-recipies between main recipies. That would be helpfull for meal planning, allowing different combinations of e.g. main dish and side dish or sauce without having to duplicate things.

**Reply by nr0dham** [reply]
Oh, now that is indeed a valid point.

**Comment by salabur** [comment]
I recently discovered kitchenowl because I am trying to move all of our services to locally hosted opensource programs. Thank you so far and I hope you continue to enjoy the project. 
Even if for me only the shopping list is "important" or realistic atm, I think that in a "serious" recipe app sooner or later there is no way around subrecipes. 

?every? Recipe should possibly be the sub-recipe of another one and be selectable as a component during recipe creation (like in the main search with the help of tag and search term). In fact, recipes should be able to contain components, which in turn have sub-components.

Example?
Schnitzel with breading{id345}(flour/egg/spices) and salad-xyz{id567}(lettuce/corn/tomatoes/salad dressing{id678}(honey/mustard/olive oil/balsamic/salt/pepper/parsley/shallot)).

Cake with whipped cream{id123}(cream/vanilla/sugar) and crème pâtissière {id234}(milk/sugar/flour/cornstarch/vanilla/yolk/butter) topped with ....
?

I will continue to test kitchenowl in the near future and then I have to try to motivate/train my wife to use something new now instead of "bring"...

**Comment by crosspolar** [comment]
I played around with the recipe-scraper Tom uses and thankfully, it's able to recognize recipe-groups:
```python
[
    IngredientGroup(
        ingredients=[
            "1 tsp active dry yeast powder",
            "1/4 cup / 65 ml warm water",
            "1 tbsp white sugar",
        ],
        purpose="Yeast Activation:",
    ),
    IngredientGroup(
        ingredients=[
            "1/2 cup /125 ml warm water",
            "4 tbsp /70g white sugar",
            "2 cups / 300g plain flour (all purpose)",
            "1 cup / 155g cornflour / cornstarch",
            "1/4 cup / 65 ml vegetable oil",
            "2 1/2 tsp baking powder",
        ],
        purpose="Dough:",
    ),
    IngredientGroup(
        ingredients=[
            "1 tbsp vegetable oil",
            "1/3 cup finely chopped escalot or white onion (Note 1)",
            "1 tbsp sugar",
            "1 1/2 tbsp soy sauce (, regular or light (not dark)",
            "1 1/2 tbsp oyster sauce (can sub Hoisin)",
            "1 tbsp sesame oil",
            "1/2 cup water",
            "1 tbsp cornflour dissolved in 1 tbsp water",
            "1 1/2 cups Chinese Barbecue Pork (, diced (Note 2)",
        ],
        purpose="Pork Filling:",
    ),
]
```

Based on this, we could continue creating sub-recipes?! At least in theory

**Comment by crosspolar** [comment]
I started working on the UI with the [chinese steamed pork buns](https://www.recipetineats.com/chinese-steamed-pork-buns/).

Ingredient-groups are straight-forward: https://excalidraw.com/#json=sloKuoOwtTpkxBTSCjLsV,J0u8YT15hlsvU2vEytE6gQ
![grafik](https://github.com/user-attachments/assets/25c9d425-7aa2-4cbe-87e3-f8e0a991af61)

Description is tougher: For the buns, you first prepare the dough and let it rest, then prepare the filling and then return to the dough. I'm not sure how to create a streamlined UI without taking hours to create a recipe: 
- sub-recipe "Dough" would surrounds sub-recipe "Filling" 
- dough and filling parts are mixed in one step when "Making Buns", i.e. "put the filling onto the dough"
- steps are not necessarily matching with recipe-groups, so merging would need a lot of manual correction

As a result, I vote for ingredient-groups but they could organized in the database as recipes, so that UI can maybe adapt as soon as we have a better idea.

**Reply by TomBursch** [reply]
Thanks for the mockup. I like the idea of recipe groups and having only one description (for now at least). The mockup looks good, the only thing I'd want to keep are ingredients without a group. It should still be possible to not use them. 

Or alternatively, what if, for a new empty recipe, ingredients and optional ingredients are the default groups. The UI could look similar to how it is now. Any group would then allow you to set if it is selected by default.

**Reply by crosspolar** [reply]
I moved on from the mockup. I think it's easier to implement if ingredient-groups are editable from the ìtem_page` below the description-part. Part of the new model would also having everything in an default-group. 

> Or alternatively, what if, for a new empty recipe, ingredients and optional ingredients are the default groups. 

You mean, optional ingredients are then part of the ingredient-group-system?

**Comment by EarMaster** [comment]
I want to bring this up again because with the new ingredient tagging in recipes I often have the problem that I need the same ingredient twice but with split amounts (let's say 50ml of cream for step 1 and 200ml of cream for step 4). Allowing to add the same ingredient multiple times would be a solution, but it would require to be able to use an index of some sort with the tagging feature.
